SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriverRPC.h

```diff
@@ -0,0 +1,241 @@
+							  Schema_ComponentData* InData);
+	static void OnUpdateWritten(TArray<UpdateToSend>& OutUpdates, Worker_EntityId EntityId, Worker_ComponentId ComponentId,
+								Schema_ComponentUpdate* InUpdate);
+
+	void FlushUpdates(TArray<UpdateToSend>&);
+	bool CanExtractRPC(Worker_EntityId);
+	bool CanExtractRPCOnServer(Worker_EntityId);
+	bool ApplyRPC(Worker_EntityId, SpatialGDK::ReceivedRPC, const FRPCMetaData& MetaData);
+
+	USpatialNetDriver* NetDriver;
+
+	USpatialLatencyTracer* LatencyTracer;
+	SpatialGDK::SpatialEventTracer* EventTracer;
+
+	TUniquePtr<SpatialGDK::RPCService> RPCService;
+	// TUniquePtr<SpatialGDK::TRPCBufferReceiver<FRPCPayload, TimestampAndETWrapper>> NetMulticastReceiver;
+};
+
+/**
+ * Server side of the RPC component.
+ * Contains client, cross server and multicast senders
+ * Contains server and cross server receivers
+ * Able to collect initial RPC data for entities to create
+ */
+UCLASS()
+class SPATIALGDK_API USpatialNetDriverServerRPC : public USpatialNetDriverRPC
+{
+	GENERATED_BODY()
+public:
+	USpatialNetDriverServerRPC();
+	void Init(USpatialNetDriver* InNetDriver, const SpatialGDK::FSubView& InActorAuthSubView,
+			  const SpatialGDK::FSubView& InActorNonAuthSubView) override;
+
+	void ProcessReceivedRPCs() override;
+	void FlushRPCUpdates() override;
+
+	TUniquePtr<SpatialGDK::TRPCQueue<FRPCPayload, FSpatialGDKSpanId>> ClientReliableQueue;
+	TUniquePtr<SpatialGDK::TRPCQueue<FRPCPayload, FSpatialGDKSpanId>> ClientUnreliableQueue;
+	TUniquePtr<SpatialGDK::TRPCQueue<FRPCPayload, FSpatialGDKSpanId>> NetMulticastQueue;
+
+protected:
+	void GetRPCComponentsOnEntityCreation(const Worker_EntityId EntityId, TArray<FWorkerComponentData>& OutData) override;
+
+	TUniquePtr<SpatialGDK::RPCBufferSender> ClientReliableSender;
+	TUniquePtr<SpatialGDK::RPCBufferSender> ClientUnreliableSender;
+	TUniquePtr<SpatialGDK::TRPCBufferReceiver<FRPCPayload, TimestampAndETWrapper>> ServerReliableReceiver;
+	TUniquePtr<SpatialGDK::TRPCBufferReceiver<FRPCPayload, TimestampAndETWrapper>> ServerUnreliableReceiver;
+};
+
+/**
+ * Client side of the RPC component.
+ * It contains server senders, and client receivers.
+ */
+UCLASS()
+class SPATIALGDK_API USpatialNetDriverClientRPC : public USpatialNetDriverRPC
+{
+	GENERATED_BODY()
+public:
+	USpatialNetDriverClientRPC();
+	void Init(USpatialNetDriver* InNetDriver, const SpatialGDK::FSubView& InActorAuthSubView,
+			  const SpatialGDK::FSubView& InActorNonAuthSubView) override;
+
+	void ProcessReceivedRPCs() override;
+	void FlushRPCUpdates() override;
+
+	TUniquePtr<SpatialGDK::TRPCQueue<FRPCPayload, FSpatialGDKSpanId>> ServerReliableQueue;
+	TUniquePtr<SpatialGDK::TRPCQueue<FRPCPayload, FSpatialGDKSpanId>> ServerUnreliableQueue;
+
+protected:
+	void GetRPCComponentsOnEntityCreation(const Worker_EntityId EntityId, TArray<FWorkerComponentData>& OutData) override;
+
+	TUniquePtr<SpatialGDK::RPCBufferSender> ServerReliableSender;
```

Comment:
Could you elaborate on why you separate the sender and queue objects?

Reply:
Someone has to "own" the sender. I could make the queue own it though.